src/charm.py

```diff
@@ -0,0 +1,595 @@
+            )
+
+        if self.config["path"] != self._stored.path:
+            prev_runner_manager = self._get_runner_manager(
+                path=str(self._stored.path)
+            )  # Casting for mypy checks.
+            if prev_runner_manager:
+                self.unit.status = MaintenanceStatus("Removing runners from old org/repo")
+                prev_runner_manager.flush()
+            self._stored.path = self.config["path"]
+
+        runner_manager = self._get_runner_manager()
+        if runner_manager:
+            self.unit.status = ActiveStatus()
+        else:
+            self.unit.status = BlockedStatus("Missing token or org/repo path config")
+
+    @catch_unexpected_charm_errors
+    def _on_update_runner_bin(self, _event: UpdateRunnerBinEvent) -> None:
+        """Handle checking update of runner binary event.
+
+        Args:
+            event: Event of checking update of runner binary.
+        """
+        runner_manager = self._get_runner_manager()
+        if not runner_manager:
+            return
+        try:
+            self.unit.status = MaintenanceStatus("Checking for runner updates")
+            runner_info = runner_manager.get_latest_runner_bin_url()
+        except urllib.error.URLError as err:
+            logger.exception("Failed to check for runner updates")
+            # Failure to download runner binary is a transient error.
+            # The charm automatically update runner binary on a schedule.
+            self.unit.status = MaintenanceStatus(f"Failed to check for runner updates: {err}")
+            return
+
+        if runner_info.download_url != self._stored.runner_bin_url:
+            self.unit.status = MaintenanceStatus("Updating runner binary")
+            try:
+                runner_manager.update_runner_bin(runner_info)
+            # Safe guard against transient unexpected error.
+            except Exception as err:  # pylint: disable=broad-exception-caught
+                logger.exception("Failed to update runner binary")
+                # Failure to download runner binary is a transient error.
+                # The charm automatically update runner binary on a schedule.
+                self.unit.status = MaintenanceStatus(f"Failed to update runner binary: {err}")
+                return
+            self._stored.runner_bin_url = runner_info.download_url
+
+            # Flush the non-busy runner and reconcile.
+            runner_manager.flush(flush_busy=False)
+            self._reconcile_runners(runner_manager)
+
+        self.unit.status = ActiveStatus()
+
+    @catch_unexpected_charm_errors
+    def _on_reconcile_runners(self, _event: ReconcileRunnersEvent) -> None:
+        """Handle the reconciliation of runners.
+
+        Args:
+            event: Event of reconciling the runner state.
+        """
+        if not RunnerManager.runner_bin_path.is_file():
+            logger.warning("Unable to reconcile due to missing runner binary")
```

Comment:
Out of curiosity, would there be a reason why this doesn't fall to BlockedStatus?

Reply:
Since the runner binary is updated/downloaded every hour by default, the charm should be able to resolve missing binary by itself.
So a better status would be Maintenance.
Might be good for the code to set to maintenance status.

Reply:
Minor comments